### DISCUSSION TITLE
Suggest variables that are only enabled during debugging

### DISCUSSION THREAD

**Discussion by zero-plusplus** [zero-plusplus/vscode-autohotkey-debug#125] [Ideas]
The debugger has the ability to write primitive values, such as strings and numbers, to variables.
Using this feature, `A_DebuggerName` is defined, which is used to determine whether you are debugging or not.

Do you have any other suggestions for variables that might be useful for debugging?

Here are my suggestions If you have any suggestions, please comment.

#### 1. `A_AllVarNames`
A list of currently available variable names. Depending on the scope, there are `A_AllVarNames`, `A_GlobalVarNames`, `A_LocalVarNames`, and `A_StaticVarNames` (H version only).

The value of this variable is a string separated by a new line "\`n", like "VarA`nVarB".

This variable can be used to output all the values of currently available variables to the log. This feature may not be used very often, but it may be useful to find the cause of bugs that are hard to reproduce.

**Comment by SAbboushi** [comment]
EXCELLENT!
I would be happy if your extension also had variables containing the values provided by ListLines/ListVars/ListHotkeys/KeyHistory (KeyHistory is the one I need and use most).
I currently use [ScriptInfo.ahk and ParseKeyHistory.ahk](https://www.autohotkey.com/boards/viewtopic.php?f=6&t=9656&p=248727&hilit=scriptinfo#p231662) to get what I need but would be great if you could provide the variables instead.  

I would also very much like to be able to access variables from other running scripts.  I've used Lexikos' DebugVars in the past to do this, but would be much better if the variables were accessible through your extension while debugging.

What do you think?

**Reply by zero-plusplus** [reply]
> I would be happy if your extension also had variables containing the values provided by ListLines/ListVars/ListHotkeys/KeyHistory (KeyHistory is the one I need and use most).

It would certainly be useful if KeyHistory could be accessed as a variable.

However, since they cannot be retrieved through the debugger, it would be technically possible to implement your suggestion, but difficult.

If implemented, each time the script breaks, need to start an AutoHotkey script to retrieve the information and write the retrieved data to a variable.
This is too costly to handle, so it is better to use ScriptInfo.ahk directly.

---

> I would also very much like to be able to access variables from other running scripts.

This one will be technically difficult. The AutoHotkey debugger can only write primitive values.

Even if it were possible, the processing cost would be high because all variables would need to be written at break time.

**Comment by SAbboushi** [comment]
OK - thanks for providing your explanations

**Comment by SAbboushi** [comment]
Lexikos has said he will consider adding native support to AHK re: ListLines/ListVars/ListHotkeys/KeyHistory so we don't need to use ScriptInfo.ahk if "[users [can] prove that the information is actually useful](https://www.autohotkey.com/boards/viewtopic.php?p=53593#p53593)".  I've already expressed my desire for him to add such native support in AHK; if you are interested too, please consider adding your comments to the post I've linked above.

**Reply by zero-plusplus** [reply]
I once needed to get a list of global variables and implemented it myself, but I forgot why I needed it.
If I can think of any useful use for it, I'll comment on it.
